pkg/api/v1/clients/wrapper/multi_watcher.go

```diff
@@ -0,0 +1,198 @@
+package wrapper
+
+import (
+	"context"
+	"sync"
+
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients"
+	"github.com/solo-io/solo-kit/pkg/api/v1/resources"
+)
+
+/*
+A wrapper.watchAggregator wraps multiple ResourceWatchers and
+aggregates a watch on each into a single Watch func
+*/
+type WatchAggregator interface {
+	clients.ResourceWatcher
+	AddWatch(w clients.ResourceWatcher) error
+	RemoveWatch(w clients.ResourceWatcher)
+}
+
+type resourceSink chan resources.ResourceList
+
+type addWatch func(watcher clients.ResourceWatcher) error
+type removeWatch func()
+
+type watchAggregator struct {
+	sources        map[clients.ResourceWatcher][]removeWatch // how to unsubscribe this watcher
+	sourcesAccess  sync.RWMutex
+	sinks          map[resourceSink]addWatch // how to subscribe the aggregator to a watcher
+	sinksAccess    sync.RWMutex
+	watchers       map[clients.ResourceWatcher]struct{}
+	watchersAccess sync.RWMutex
+}
+
+func NewWatchAggregator() WatchAggregator {
+	sources := make(map[clients.ResourceWatcher][]removeWatch)
+	sinks := make(map[resourceSink]addWatch)
+	watchers := make(map[clients.ResourceWatcher]struct{})
+	return &watchAggregator{sources: sources, sinks: sinks, watchers: watchers}
+}
+
+func (c *watchAggregator) Watch(namespace string, opts clients.WatchOpts) (<-chan resources.ResourceList, <-chan error, error) {
+	opts = opts.WithDefaults()
+	// create new sinks for this watch
+	out := make(chan resources.ResourceList)
+	aggregatedErrs := make(chan error)
+
+	// a shared map that will be used to merge resources from different watchers
+	listsByWatcher := make(resourcesByWatcher)
+	access := sync.RWMutex{}
+
+	// create a wait group for sources
+	// so we can wait for all sources watches to close
+	// before closing the sink channel (when this watch is canceled)
+	sourceWatches := sync.WaitGroup{}
+
+	// construct a func for adding an input watcher to this sink
+	addWatch := func(watcher clients.ResourceWatcher) (err error) {
+		sourceWatches.Add(1)
+
+		// this function starts a watch for the watcher using the root context
+		// we want to cancel it if :
+		// the root context is cancelled
+		// the watcher is removed
+		ctx, cancel := context.WithCancel(opts.Ctx)
+
+		// start a watch for the watcher on this namespace
+		source, errs, err := watcher.Watch(namespace, clients.WatchOpts{
+			Ctx:         ctx,
+			RefreshRate: opts.RefreshRate,
+			Selector:    opts.Selector,
+		})
+		if err != nil {
+			return err
+		}
+
+		// read lists from the source channel,
+		// group its resources by type
+		go func() {
+			defer cancel()
+			defer sourceWatches.Done()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case err := <-errs:
+					// if the source starts returning errors, remove its list from the snasphot
```

Comment:
is it possible from a source to recover from the error?
or is the semantics is that error is the last message?

Reply:
it can recover if we receive a new list, it gets added back to the snapshot